api/client/contextdialer.go

```diff
@@ -78,20 +79,65 @@ func newTunnelDialer(ssh ssh.ClientConfig, keepAlivePeriod, dialTimeout time.Dur
 			return nil, trace.Wrap(err)
 		}
 
-		ssh.Timeout = dialTimeout
-		sconn, err := sshutils.NewClientConnWithDeadline(conn, addr, &ssh)
+		sconn, err := sshConnect(conn, ssh, dialTimeout, addr)
 		if err != nil {
-			return nil, trace.NewAggregate(err, conn.Close())
+			return nil, trace.Wrap(err)
 		}
+		return sconn, nil
+	})
+}
 
-		// Build a net.Conn over the tunnel. Make this an exclusive connection:
-		// close the net.Conn as well as the channel upon close.
-		conn, _, err = sshutils.ConnectProxyTransport(sconn.Conn, &sshutils.DialReq{
-			Address: constants.RemoteAuthServer,
-		}, true)
+// newTLSRoutingTunnelDialer makes a reverse tunnel TLS Routing dialer to connect to an Auth server
+// through the SSH reverse tunnel on the proxy.
+func newTLSRoutingTunnelDialer(ssh ssh.ClientConfig, keepAlivePeriod, dialTimeout time.Duration, discoveryAddr string, insecure bool) ContextDialer {
+	return ContextDialerFunc(func(ctx context.Context, network, addr string) (conn net.Conn, err error) {
+		tunnelAddr, err := webclient.GetTunnelAddr(ctx, discoveryAddr, insecure, nil)
 		if err != nil {
-			return nil, trace.NewAggregate(err, sconn.Close())
+			return nil, trace.Wrap(err)
 		}
-		return conn, nil
+		dialer := &net.Dialer{
+			Timeout:   dialTimeout,
+			KeepAlive: keepAlivePeriod,
+		}
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		host, err := webclient.ExtractHost(tunnelAddr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		tlsConn, err := tls.DialWithDialer(dialer, network, tunnelAddr, &tls.Config{
+			NextProtos:         []string{constants.ALPNSNIProtocolReverseTunnel},
+			InsecureSkipVerify: insecure,
+			ServerName:         host,
+		})
```

Comment:
nit: `tls.DialWithDialer` dials with `context.Background()`. ideally we can dial with the passed in `ctx`, by wrapping `net.Dialer` in a `tls.Dialer` and call its`DialContext`

Reply:
Updated, thanks for suggestion.